syntax = "proto3";

message ShardData {
  int32 mapper_id = 1;
  string shard_file = 2;
  int32 start = 3;
  int32 end = 4;
  repeated Centroid centroids = 5;
  int32 R = 6;
}

message Centroid {
  int32 centroid_id = 1;
  double x = 2;
  double y = 3;
}

message SendDataRequest {
    int32 reducer_id = 1;
}

message MapperDataResponse {
    repeated ReduceData data = 1;
}

// Define a message that represents a single piece of data to be reduced.
message ReduceData {
    string key = 1;
    string value = 2;
}


message MapperResponse {
  string result = 1;
}

service Mapper {
  rpc MapData(ShardData) returns (MapperResponse);
  rpc Mapper2ReduceData(SendDataRequest) returns (MapperDataResponse);
}